cl/assets/react/TagSelect.tsx

```diff
@@ -0,0 +1,254 @@
+import React from 'react';
+import { useVirtual } from 'react-virtual';
+import { useCombobox } from 'downshift';
+import { ListItem } from './ListItem';
+import { useTags } from './_useTags';
+import { Association, UserState } from './_types';
+
+
+const TagSelect: React.FC<UserState> = ({ userId, userName, editUrl, docket}) => {
+  const [validationError, setValidationError] = React.useState<null | string>(null);
+
+  const {
+    infiniteQueryState: { status, canFetchMore, isFetching, isFetchingMore, fetchMore },
+    textVal,
+    setTextVal,
+    tags,
+    associations,
+    addNewTag,
+    addNewAssociation,
+    deleteAssociation,
+  } = useTags({ docket: docket as number, enabled: !!docket && userName, userId: userId });
+
+  const parentRef = React.useRef(null);
+  const rowVirtualizer = useVirtual({
+    size: canFetchMore ? tags.length + 1 : tags.length,
+    parentRef,
+    estimateSize: React.useCallback(() => 40, []),
+  });
+
+  // fetchmore if we are at the bottom of the list
+  React.useEffect(() => {
+    const [lastItem] = [...rowVirtualizer.virtualItems].reverse();
+
+    if (!lastItem) return;
+
+    if (lastItem.index === rowVirtualizer.virtualItems.length - 1 && canFetchMore && !isFetchingMore) {
+      console.log('fetching more');
+      fetchMore();
+    }
+  }, [canFetchMore, fetchMore, tags.length, isFetchingMore, rowVirtualizer.virtualItems]);
+
+  const {
+    isOpen,
+    getToggleButtonProps,
+    getLabelProps,
+    getMenuProps,
+    getInputProps,
+    getComboboxProps,
+    highlightedIndex,
+    getItemProps,
+  } = useCombobox({
+    inputValue: textVal,
+    itemToString: (item) => (item ? item.name : ''),
+    // set to none to select multiple
+    selectedItem: null,
+    items: tags,
+    scrollIntoView: () => {},
+    stateReducer: (state, actionAndChanges) => {
+      const { changes, type } = actionAndChanges;
+      switch (type) {
+        case useCombobox.stateChangeTypes.InputKeyDownEnter:
+          return {
+            ...changes,
+            isOpen: true,
+            highlightedIndex: state.highlightedIndex,
+            inputValue: 'return'
+          }
+        case useCombobox.stateChangeTypes.ItemClick:
+          return {
+            ...changes,
+            isOpen: true, // keep menu open after selection.
+            highlightedIndex: state.highlightedIndex,
+            inputValue: '',
+          };
+        default:
+          return changes;
+      }
+    },
+    onInputValueChange: ({inputValue}) => {
+      if (inputValue !== 'return') return;
+      let d3i0 = document.querySelector("div.list-group > div > div > div > div")!.textContent || ""
```

Comment:
Can you please do something better here? Using this selector is fragile, and the variable name is unclear.

Reply:
yes.  yes it is ugly

Reply:
√